src/net/p2p.rs

```diff
@@ -3613,2219 +3750,4119 @@ impl PeerNetwork {
+                            return (false, faulty_peers, None);
+                        }
+                        Ok((false, Some(_))) => {
+                            // should never happen
+                            warn!("Reached invalid state in microblock tip sync protocol when waiting for microblocks, resetting...");
+                            self.microblock_tip_sync_reset();
+                            return (true, faulty_peers, None);
+                        }
+                        Err(e) => {
+                            warn!(
+                                "{:?}: microblock_tip_sync_recv_microblock_response failed {:?}",
+                                &self.local_peer, &e
+                            );
+                            self.microblock_tip_sync_reset();
+                            return (true, faulty_peers, None);
+                        }
+                    }
                 }
             }
-            msgs.push(msg);
-            debug!(
-                "{:?}: Event {} has {} messages buffered",
-                &self.local_peer,
-                event_id,
-                msgs.len()
-            );
-        } else {
-            self.pending_messages.insert(event_id, vec![msg]);
-            debug!(
-                "{:?}: Event {} has 1 messages buffered",
-                &self.local_peer, event_id
-            );
-        }
-    }
-
-    /// Do we need a block or microblock stream, given its sortition's consensus hash?
-    fn need_block_or_microblock_stream(
-        sortdb: &SortitionDB,
-        chainstate: &StacksChainState,
-        consensus_hash: &ConsensusHash,
-        is_microblock: bool,
-    ) -> Result<bool, net_error> {
-        let sn = SortitionDB::get_block_snapshot_consensus(sortdb.conn(), &consensus_hash)?
-            .ok_or(chainstate_error::NoSuchBlockError)?;
-        let block_hash_opt = if sn.sortition {
-            Some(sn.winning_stacks_block_hash)
-        } else {
-            None
-        };
-
-        let inv = chainstate.get_blocks_inventory(&[(consensus_hash.clone(), block_hash_opt)])?;
-        if is_microblock {
-            // checking for microblock absence
-            Ok(inv.microblocks_bitvec[0] == 0)
-        } else {
-            // checking for block absence
-            Ok(inv.block_bitvec[0] == 0)
         }
     }
 
-    /// Handle unsolicited BlocksAvailable.
-    /// Update our inv for this peer.
-    /// Mask errors.
-    /// Return whether or not we need to buffer this message
-    fn handle_unsolicited_BlocksAvailable(
+    /// Do the actual work in the state machine.
+    /// Return true if we need to prune connections.
+    fn do_network_work(
         &mut self,
         sortdb: &SortitionDB,
-        chainstate: &StacksChainState,
-        event_id: usize,
-        new_blocks: &BlocksAvailableData,
+        mempool: &MemPoolDB,
+        chainstate: &mut StacksChainState,
```

Comment:
Why is this `&mut` now?

Reply:
I believe this is just strange formatting on the diff. My PR does not change the type of `chainstate` for either the functions `do_network_work` or `handle_unsolicited_BlocksAvailable` (as compared to the branch `develop`)